File: app/src/main/java/com/cameronbanga/skyscraper/viewmodels/ChatViewModel.kt
package com.cameronbanga.skyscraper.viewmodels

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.cameronbanga.skyscraper.models.ConvoView
import com.cameronbanga.skyscraper.models.MessageInput
import com.cameronbanga.skyscraper.models.MessageUnion
import com.cameronbanga.skyscraper.models.MessageView
import com.cameronbanga.skyscraper.services.ATProtoClient
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

/**
 * ViewModel for managing an individual chat conversation
 * Handles loading messages, sending messages, and polling for updates
 */
class ChatViewModel(
    application: Application,
    val conversation: ConvoView
) : AndroidViewModel(application) {

    private val client = ATProtoClient.getInstance()

    private val _messages = MutableStateFlow<List<MessageUnion>>(emptyList())
    val messages: StateFlow<List<MessageUnion>> = _messages.asStateFlow()

    private val _messageText = MutableStateFlow("")
    val messageText: StateFlow<String> = _messageText.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _isSending = MutableStateFlow(false)
    val isSending: StateFlow<Boolean> = _isSending.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private var cursor: String? = null
    private var pollingJob: Job? = null

    override fun onCleared() {
        super.onCleared()
        stopPolling()
    }

    fun startPolling() {
        // Cancel any existing polling task
        pollingJob?.cancel()

        pollingJob = viewModelScope.launch {
            while (isActive) {
                // Wait 2.5 seconds
                delay(2500)

                if (isActive) {
                    pollForNewMessages()
                }
            }
        }
    }

    fun stopPolling() {
        pollingJob?.cancel()
        pollingJob = null
    }

    fun loadMessages() {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null

            try {
                val response = client.getMessages(conversation.id, limit = 50)
                _messages.value = response.messages.reversed() // Reverse to show oldest first
                cursor = response.cursor

                // Mark as read
                try {
                    client.updateRead(conversation.id)
                } catch (e: Exception) {
                    // Ignore read update errors
                }
            } catch (e: Exception) {
                _errorMessage.value = e.message ?: "Failed to load messages"
            } finally {
                _isLoading.value = false
            }
        }
    }

    fun updateMessageText(text: String) {
        _messageText.value = text
    }

    fun sendMessage() {
        val textToSend = _messageText.value.trim()
        if (textToSend.isEmpty()) {
            return
        }

        viewModelScope.launch {
            _messageText.value = "" // Clear immediately for better UX
            _isSending.value = true

            try {
                val messageInput = MessageInput(text = textToSend)
                val response = client.sendMessage(conversation.id, messageInput)

                // Add the new message to the list
                val newMessage = MessageView(
                    id = response.id,
                    rev = response.rev,
                    text = response.text,
                    facets = response.facets,
                    embed = response.embed,
                    sender = response.sender,
                    sentAt = response.sentAt
                )
                _messages.value = _messages.value + MessageUnion.MessageView(newMessage)
            } catch (e: Exception) {
                // Restore the text so user can try again
                _messageText.value = textToSend
                _errorMessage.value = "Failed to send message. Please try again."
            } finally {
                _isSending.value = false
            }
        }
    }

    fun loadMoreMessages() {
        val currentCursor = cursor
        if (currentCursor == null || _isLoading.value) {
            return
        }

        viewModelScope.launch {
            try {
                val response = client.getMessages(conversation.id, limit = 50, cursor = currentCursor)
                _messages.value = response.messages.reversed() + _messages.value
                cursor = response.cursor
            } catch (e: Exception) {
                // Ignore pagination errors
            }
        }
    }

    private suspend fun pollForNewMessages() {
        if (_isLoading.value || _isSending.value) {
            return
        }

        try {
            // Get the latest messages without cursor to fetch newest ones
            val response = client.getMessages(conversation.id, limit = 50)
            val newMessages = response.messages.reversed()

            // Find messages we don't have yet
            val existingIds = _messages.value.map { it.id }.toSet()
            val messagesToAdd = newMessages.filter { !existingIds.contains(it.id) }

            if (messagesToAdd.isNotEmpty()) {
                _messages.value = _messages.value + messagesToAdd

                // Mark as read
                try {
                    client.updateRead(conversation.id)
                } catch (e: Exception) {
                    // Ignore read update errors
                }
            }
        } catch (e: Exception) {
            // Ignore polling errors
        }
    }

    fun clearError() {
        _errorMessage.value = null
    }
}

File: Skyscraper-iOS-20251106/Skyscraper/ViewModels/ChatViewModel.swift
//
//  ChatViewModel.swift
//  Skyscraper
//
//  Manages individual chat conversation
//

import Foundation
import Combine

@MainActor
class ChatViewModel: ObservableObject {
    @Published var messages: [MessageUnion] = []
    @Published var messageText = ""
    @Published var isLoading = false
    @Published var isSending = false
    @Published var errorMessage: String?

    let conversation: ConvoView
    private let client = ATProtoClient.shared
    private var cursor: String?
    private var pollingTask: Task<Void, Never>?

    init(conversation: ConvoView) {
        self.conversation = conversation
    }

    deinit {
        pollingTask?.cancel()
    }

    func startPolling() {
        // Cancel any existing polling task
        pollingTask?.cancel()

        pollingTask = Task {
            while !Task.isCancelled {
                // Wait 2.5 seconds
                try? await Task.sleep(nanoseconds: 2_500_000_000)

                guard !Task.isCancelled else { break }

                // Poll for new messages
                await pollForNewMessages()
            }
        }
    }

    func stopPolling() {
        pollingTask?.cancel()
        pollingTask = nil
    }

    func loadMessages() async {
        isLoading = true
        errorMessage = nil

        do {
            let response = try await client.getMessages(convoId: conversation.id, limit: 50)
            messages = response.messages.reversed() // Reverse to show oldest first
            cursor = response.cursor
            print("✅ Loaded \(messages.count) messages")

            // Mark as read
            try? await client.updateRead(convoId: conversation.id)
        } catch {
            errorMessage = error.localizedDescription
            print("❌ Failed to load messages: \(error)")
        }

        isLoading = false
    }

    func sendMessage() async {
        guard !messageText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return
        }

        let textToSend = messageText
        messageText = "" // Clear immediately for better UX

        isSending = true

        do {
            let messageInput = MessageInput(text: textToSend)
            let response = try await client.sendMessage(convoId: conversation.id, message: messageInput)

            // Add the new message to the list
            let newMessage = MessageView(
                id: response.id,
                rev: response.rev,
                text: response.text,
                facets: response.facets,
                embed: response.embed,
                sender: response.sender,
                sentAt: response.sentAt
            )
            messages.append(.messageView(newMessage))

            print("✅ Message sent successfully")
        } catch {
            print("❌ Failed to send message: \(error)")
            // Restore the text so user can try again
            messageText = textToSend
            errorMessage = "Failed to send message. Please try again."
        }

        isSending = false
    }

    func loadMoreMessages() async {
        guard let cursor = cursor, !isLoading else {
            return
        }

        do {
            let response = try await client.getMessages(convoId: conversation.id, limit: 50, cursor: cursor)
            messages.insert(contentsOf: response.messages.reversed(), at: 0)
            self.cursor = response.cursor
        } catch {
            print("❌ Failed to load more messages: \(error)")
        }
    }

    private func pollForNewMessages() async {
        guard !isLoading && !isSending else {
            return
        }

        do {
            // Get the latest messages without cursor to fetch newest ones
            let response = try await client.getMessages(convoId: conversation.id, limit: 50)
            let newMessages = response.messages.reversed()

            // Find messages we don't have yet
            let existingIds = Set(messages.map { $0.id })
            let messagesToAdd = newMessages.filter { !existingIds.contains($0.id) }

            if !messagesToAdd.isEmpty {
                messages.append(contentsOf: messagesToAdd)
                print("✅ Polled and found \(messagesToAdd.count) new message(s)")

                // Mark as read
                try? await client.updateRead(convoId: conversation.id)
            }
        } catch {
            print("❌ Failed to poll for new messages: \(error)")
        }
    }
}
